### PR TITLE
Create Type Definitions for react-visibility-sensor

### PR DESCRIPTION
**Pull request by JRasmusBm**
Adding new definitions for the react-visibility-sensor on npm. Refer to https://github.com/joshwnj/react-visibility-sensor/issues/112. This is my first time contributing to DefinitelyTyped (and probably not last), so any feedback would be greatly appreciated.

- [ x] Use a meaningful title for the pull request. Include the name of the package modified.
- [ x] Test the change in your own code. (Compile and run.)
- [ x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [ x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [ x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Adding a new definition:
- [x ] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x ] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x ] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

**Comment by tup1tsa** [comment]
Why do you need 15 commits for 80 lines of code? Use merge-squash

File: types/react-visibility-sensor/index.d.ts [hunk 1]
```diff
@@ -1,5 +1,5 @@
 // Type definitions for react-visibility-sensor 3.11
-// Project: https://github.com/joshwnj/react-visibility-sensor
+// Project: https://github.com/joshwnj/react-visibility-sensor#readme
```
Comment (intolerance) [inline types/react-visibility-sensor/index.d.ts]: Why did you add `#readme` to this?

**Review by intolerance — CHANGES_REQUESTED** [review]
Please provide feedback as requested in the other comments.

File: types/react-visibility-sensor/index.d.ts (hunk 1 above)
Comment (JRasmusBm) [reply types/react-visibility-sensor/index.d.ts]: I first made the project based on an existing project. Then when I was done and about to submit I saw the check-box,

`[x ] Create it with dts-gen --dt, not by basing it on an existing project.`

I did that and diffed it with my version. The only difference was how the link in the header looked and the ordering of arguments in `tsconfig.json`. In the generated files it had the `#readme` at the end of the link, so I added it to conform with that template.

**Comment by JRasmusBm** [comment]
Wait with merging!
There seems to be a bug in the types!

**Review by intolerance — APPROVED** [review]
Thanks for updating the `Shape`.

File: types/react-visibility-sensor/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  resizeThrottle?: number;
+  delayedCall?: boolean;
+  children?: (
+    args: { isVisible: boolean; visibilityRect?: Shape }
+  ) => React.ReactNode;
+}
+
+declare const ReactVisibilitySensor: React.StatelessComponent<Props>;
+
+export default ReactVisibilitySensor;
```
Comment (armanio123) [inline types/react-visibility-sensor/index.d.ts]: It seems like the library not using `export default`, so this .d.ts file should not either.

Instead, you should follow the appropriate template from [the TypeScript handbook](http://www.typescriptlang.org/docs/handbook/declaration-files/templates.html).

File: types/react-visibility-sensor/react-visibility-sensor-tests.tsx [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import * as React from "react";
+
+import ReactVisibilitySensor, { Shape } from "react-visibility-sensor";
```
Comment (armanio123) [inline types/react-visibility-sensor/react-visibility-sensor-tests.tsx]: This will need to change as well because of the `default`

**Review by armanio123 — CHANGES_REQUESTED** [review]

File: types/react-visibility-sensor/index.d.ts (hunk 2 above)
Comment (JRasmusBm) [reply types/react-visibility-sensor/index.d.ts]: Thank you for the time to review my pull request. Is this how you meant?

**Comment by JRasmusBm** [comment]
@armanio123 Thank you for the time to review the pull-request. I made the changes you asked for (I hope, otherwise please clarify further), which sadly meant I had to make void the changes suggested by @intolerance.

**Review by armanio123 — APPROVED** [review]
